src/encoders.js

```diff
@@ -0,0 +1,111 @@
+/** @typedef {string | Buffer | ArrayBuffer | Buffer[]} Data */
+/** @typedef {(data: Uint8Array | ArrayBuffer) => Promise<Data>} Encoder */
+/** @typedef {(data: Data) => Uint8Array} Decoder */
+
+/**
+ * @typedef {Object} EncoderSet
+ * @property {Encoder} requestEncoder
+ * @property {Decoder} responseDecoder
+ */
+
+/**
+ * @type {EncoderSet}
+ */
+const defaultEncoderSet = {
+  requestEncoder: async (data) => {
+    return data
+  },
+  responseDecoder: (data) => {
+    if (data instanceof Uint8Array) {
+      return data
+    } else if (data instanceof ArrayBuffer) {
+      return new Uint8Array(data)
+    } else if (data instanceof Buffer) {
+      return new Uint8Array(data)
+    } else {
+      throw new TypeError('Unexpected data type')
+    }
+  }
+}
+
+/**
+ * @type {EncoderSet}
+ */
+const awsEncoderSetForBrowser = {
+  requestEncoder: async (data) => {
+    const base64Url = await new Promise((resolve, reject) => {
+      const reader = new FileReader() // eslint-disable-line
+      reader.onload = () => resolve(reader.result)
+      reader.onerror = () => reject(reader.error)
+
+      reader.readAsDataURL(new Blob([data])) // eslint-disable-line
+    })
+
+    /*
+     * The result looks like
+     * "data:application/octet-stream;base64,<your base64 data>",
+     * so we split off the beginning:
+     */
+    return base64Url.substring(base64Url.indexOf(',') + 1)
+  },
+  /** @param {any} data */
+  responseDecoder: (data) => {
+    let arr = new Uint8Array(data)
```

Comment:
maybe this can also be simplified so we don't double assign

Reply:
Example? I pulled this from your original code, but it does feel awkward